### ISSUE TITLE
Issues with submodule processing

### ISSUE THREAD

**Issue by directhex** [flatpak/flatpak#216]
I'm trying to switch a source from archive to git, but am getting errors on the submodules:

```
Fetching git repo git://github.com/mono/mono.git
Fetching git repo git://github.com/mono/Lucene.Net.Light.git
Fetching git repo git://github.com/mono/Newtonsoft.Json.git
Fetching git repo git://github.com/mono/aspnetwebstack.git
Failed to download sources: module mono: Child process exited with code 1
```

This comes from the following flatpak-builder json:

```
        {
            "name": "mono",
            "config-opts": ["--disable-boehm","--with-monodroid=yes"],
            "no-parallel-make": true,
            "sources": [
                {
                    "type": "git",
                    "url": "git://github.com/mono/mono.git",
                    "branch": "mono-4.4.2.8"
                },
                {
                    "type": "shell",
                    "commands": [ "./autogen.sh",
                                  "make get-monolite-latest"
                    ]
                }
            ],
            "post-install": [
                "cert-sync /etc/ssl/certs/ca-certificates.crt"
            ]
        },
```

It seems to be a flatpak-specific issue - normal git is happy

**Comment by matthiasclasen** [comment]
Seems to work for me. Here is what I get when I add this fragment to a full manifest:

Downloading sources
Cloning git repo git://github.com/mono/mono.git
Cloning into bare repository 'git_github.com_mono_mono.git.clone_tmp'...
remote: Counting objects: 1124350, done.
remote: Compressing objects: 100% (529/529), done.
Rceiving objects: ...

**Comment by directhex** [comment]
It's not touching the submodules at all for you?

Maybe it's related to the version of git on the host OS?

**Comment by matthiasclasen** [comment]
submodules ? Could you show a complete json file to try ?

**Comment by bochecha** [comment]
I'm having the same issue with the [org.gnome.Games](https://git.gnome.org/browse/gnome-apps-nightly/tree/org.gnome.Games.json) manifest.

```
$ flatpak-builder appdir org.gnome.Games.json 
Downloading sources
Cloning git repo https://git.gnome.org/browse/retro-gtk
Cloning into bare repository 'https_git.gnome.org_browse_retro-gtk.clone_tmp'...
remote: Counting objects: 2956, done.
remote: Compressing objects: 100% (737/737), done.
remote: Total 2956 (delta 2209), reused 2938 (delta 2198)
Receiving objects: 100% (2956/2956), 731.02 KiB | 523.00 KiB/s, done.
Resolving deltas: 100% (2209/2209), done.
Checking connectivity... done.
Cloning git repo https://git.gnome.org/browse/retro-plugins
Cloning into bare repository 'https_git.gnome.org_browse_retro-plugins.clone_tmp'...
remote: Counting objects: 193, done.
remote: Compressing objects: 100% (105/105), done.
remote: Total 193 (delta 109), reused 153 (delta 84)
Receiving objects: 100% (193/193), 37.26 KiB | 0 bytes/s, done.
Resolving deltas: 100% (109/109), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/stella-libretro.git
Cloning into bare repository 'https_github.com_libretro_stella-libretro.git.clone_tmp'...
remote: Counting objects: 1719, done.
remote: Compressing objects: 100% (61/61), done.
remote: Total 1719 (delta 36), reused 0 (delta 0), pack-reused 1657
Receiving objects: 100% (1719/1719), 3.89 MiB | 1.73 MiB/s, done.
Resolving deltas: 100% (731/731), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/prosystem-libretro.git
Cloning into bare repository 'https_github.com_libretro_prosystem-libretro.git.clone_tmp'...
remote: Counting objects: 569, done.
remote: Total 569 (delta 0), reused 0 (delta 0), pack-reused 569
Receiving objects: 100% (569/569), 273.14 KiB | 0 bytes/s, done.
Resolving deltas: 100% (402/402), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/gambatte-libretro.git
Cloning into bare repository 'https_github.com_libretro_gambatte-libretro.git.clone_tmp'...
remote: Counting objects: 10456, done.
remote: Total 10456 (delta 0), reused 0 (delta 0), pack-reused 10456
Receiving objects: 100% (10456/10456), 2.84 MiB | 1007.00 KiB/s, done.
Resolving deltas: 100% (8349/8349), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/mame.git
Cloning into bare repository 'https_github.com_libretro_mame.git.clone_tmp'...
remote: Counting objects: 664395, done.
remote: Total 664395 (delta 0), reused 0 (delta 0), pack-reused 664395
Receiving objects: 100% (664395/664395), 491.61 MiB | 5.97 MiB/s, done.
Resolving deltas: 100% (567738/567738), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/beetle-ngp-libretro.git
Cloning into bare repository 'https_github.com_libretro_beetle-ngp-libretro.git.clone_tmp'...
remote: Counting objects: 6168, done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 6168 (delta 5), reused 0 (delta 0), pack-reused 6156
Receiving objects: 100% (6168/6168), 23.39 MiB | 7.15 MiB/s, done.
Resolving deltas: 100% (3396/3396), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/nestopia.git
Cloning into bare repository 'https_github.com_libretro_nestopia.git.clone_tmp'...
remote: Counting objects: 5253, done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 5253 (delta 2), reused 0 (delta 0), pack-reused 5245
Receiving objects: 100% (5253/5253), 4.04 MiB | 2.81 MiB/s, done.
Resolving deltas: 100% (3721/3721), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/beetle-pce-fast-libretro.git
Cloning into bare repository 'https_github.com_libretro_beetle-pce-fast-libretro.git.clone_tmp'...
remote: Counting objects: 5811, done.
remote: Compressing objects: 100% (50/50), done.
remote: Total 5811 (delta 17), reused 0 (delta 0), pack-reused 5761
Receiving objects: 100% (5811/5811), 23.01 MiB | 7.15 MiB/s, done.
Resolving deltas: 100% (3196/3196), done.
Checking connectivity... done.
Cloning git repo https://github.com/libretro/pcsx_rearmed.git
Cloning into bare repository 'https_github.com_libretro_pcsx_rearmed.git.clone_tmp'...
remote: Counting objects: 7959, done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 7959 (delta 2), reused 0 (delta 0), pack-reused 7952
Receiving objects: 100% (7959/7959), 4.85 MiB | 3.38 MiB/s, done.
Resolving deltas: 100% (5381/5381), done.
Checking connectivity... done.
Failed to download sources: module retro-plugins: Child process exited with code 1
```

The issue also happens when `flatpak-builder` downloads the sources for the submodules in `retro-plugins`.

It seems to have downloaded them all, but failed at the last one.

I can clone it manually with `git` just fine, though.

**Comment by bochecha** [comment]
Looking into it a bit more, `retro-plugins` has various submodules, and the one that causes the failure with `flatpak-builder` is defined as follows:

```ini
[submodule "plugins/playstation/pcsx_rearmed"]
       path = plugins/playstation/pcsx_rearmed
       url = https://github.com/libretro/pcsx_rearmed.git
```

It seems perfectly correct, similar to the other submodule definitions in `retro-plugins`. The only difference is that this is the only submodule with an `_` character. I doubt that could be the problem, though...

In any case, removing that submodule from `retro-plugins`, I can finally build GNOME Games with `flatpak-builder`.

**Comment by bochecha** [comment]
So I can reproduce the problem with flatpak from master.

And I think I found the problem.

`flatpak-builder` runs `git submodule status` to get the list of submodules. In the case of `retro-plugins`, the output I get running that command is:

```
$ git submodule status
-7b62cff4d070ff51668c3c434a33c3b2d7c3e7d0 plugins/atari-2600/stella-libretro
-fe7530c9586dad56e400791a9a353c2d27e6d7aa plugins/atari-7800/prosystem-libretro
-9df855665f43c96944eff3085a082aad12be6723 plugins/game-boy/gambatte-libretro
-bb828fa63768b4aea7b535a473fad3024a40220c plugins/mame/mame
-3760b9203ed907102924822044e0f9174253db74 plugins/neo-geo-pocket/beetle-ngp-libretro
-845bcea7e0ad47f8d66046712353206a53ddeb94 plugins/nes/nestopia
-097458df2a1535bffb5df56e6517c03773b39022 plugins/pc-engine/beetle-pce-fast-libretro
-3e2464bab1577295b8f4eee9756dbaf22a03ea67 plugins/playstation/pcsx_rearmed
-9a18ef0125a68a18ca4c0512216db07a22ac0486 plugins/snes/bsnes-libretro
```

Then `flatpak-builder` iterates over all the lines in this output. So I started printing each line, and I'm getting this output from `flatpak-builder`:

```
Downloading sources
Fetching git repo https://git.gnome.org/browse/retro-gtk
Fetching git repo https://git.gnome.org/browse/retro-plugins
Line: [-7b62cff4d070ff51668c3c434a33c3b2d7c3e7d0 plugins/atari-2600/stella-libretro]
Fetching git repo https://github.com/libretro/stella-libretro.git
Line: [-fe7530c9586dad56e400791a9a353c2d27e6d7aa plugins/atari-7800/prosystem-libretro]
Fetching git repo https://github.com/libretro/prosystem-libretro.git
Line: [-9df855665f43c96944eff3085a082aad12be6723 plugins/game-boy/gambatte-libretro]
Fetching git repo https://github.com/libretro/gambatte-libretro.git
Line: [-bb828fa63768b4aea7b535a473fad3024a40220c plugins/mame/mame]
Fetching git repo https://github.com/libretro/mame.git
Line: [-3760b9203ed907102924822044e0f9174253db74 plugins/neo-geo-pocket/beetle-ngp-libretro]
Fetching git repo https://github.com/libretro/beetle-ngp-libretro.git
Line: [-845bcea7e0ad47f8d66046712353206a53ddeb94 plugins/nes/nestopia]
Fetching git repo https://github.com/libretro/nestopia.git
Line: [-097458df2a1535bffb5df56e6517c03773b39022 plugins/pc-engine/beetle-pce-fast-libretro]
Fetching git repo https://github.com/libretro/beetle-pce-fast-libretro.git
Line: [-3e2464bab1577295b8f4eee9756dbaf22a03ea67 plugins/playstation/pcsx_rearmed]
Fetching git repo https://github.com/libretro/pcsx_rearmed.git
Line: [-21604a047941b8fe81d381ede0371c75da964afd frontend/libpicofe]
Failed to download sources: module retro-plugins: Child process exited with code 1
```

And there we have the failing line: that `frontend/libpicode` is not a submodule of `retro-plugins` at all!

However, it is a submodule of `plugins/playstation/pcsx_rearmed`.

So it seems that the problem happens when recursively going down the submodules of submodules.

**Comment by bochecha** [comment]
So to sum up, `flatpak-builder` assumes that the output of the `git submodule status` gives it the name of the config option for a submodule, when in fact it returns the `path`.

That only works if the submodules use their `path` as their section name, but I don't think `flatpak-builder` should make that assumption.

**Comment by bochecha** [comment]
@directhex Can you test building your app with the code from my pull request? It fixes the issue for me. :smile:

**Comment by directhex** [comment]
Sorry, I never got around to testing when this PR was still open

To confirm, it is fixed in the 0.6.9 release, which includes your changes.
